### PR TITLE
Refactoring `procBlock`

### PR DESCRIPTION
**Pull request by wodeni**
Refactored `procBlock` and some aliasing for readability of `Style`
* added underscore to all internal ids for labels
* modified tree.sty due to changes in label definition
* separated out selector logic.
* different `procBlock` cases
    * case 1: just one `global` selector -> process objs and constrs
once
    * case 2: just one selector -> the original logic
    * case 3: a list of comma separated selectors -> objs and constrs only and compute valid combinations among the selctors

### CODE REVIEW

File: src/Style.hs [hunk 1]
```diff
@@ -283,12 +269,40 @@ attribute = identifier -- TODO: Naming convention - same as identifiers?
...
+-- Translates from a Style AST to:
+-- * a list of geometric primitives
+-- * a list of constraints
+-- * a list of objectives
+
+-- Type aliases for readability in this section
+-- | 'StyContext' maintains the current output of the translater
+type StyContext a =
+    (StyDict, -- | dictionary mapping Substance ID to Style objects
+    [(ObjFnOn a, Weight a, [Name], [a])],    -- | List of objective functions
```
Comment (hypotext) [inline src/Style.hs]: Why not use `ObjFnInfo a` and `ConstrFnInfo a`?

Comment (wodeni) [reply src/Style.hs]: Oh true. Pushing a quick fix now. Any other suggestion?

File: src/Style.hs [hunk 2]
```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
...
+            let newSpec = foldl (procAssign varmap) spec statements
+            in M.insert (spId newSpec) newSpec dct
+
+-- Case 3: multiple selectors, which in the current Style languages means
+-- we would take a subset of the cartesian product among the selectors.
+-- TEMP: the system does not support assignment statements in these blocks,
+-- because it is not clear that which Substance object the shapes associate with
+procBlock (dict, objFns, constrFns) (selectors, stmts) =
+    let mapsAndSpecs  = dict `matchWithAll` selectors
+        mergedMaps    = mergeMaps $ map (map fst) $ mapsAndSpecs
```
Comment (hypotext) [inline src/Style.hs]: $ not needed

File: src/Style.hs [hunk 3]
```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
...
+        newObjFns     = concatMap (procObjFn identityMap [])    stmts
+        newConstrFns  = concatMap (procConstrFn identityMap []) stmts
+    in (dict, objFns ++ newObjFns, constrFns ++ newConstrFns)
+
+-- Case 2: only one selector, no need to compute combinations.
+-- this function will process: (1) objectives, (2) constraints and (3)
+-- assignments
+procBlock (dict, objFns, constrFns) (selector : [], stmts) =
+    let mapsAndSpecs  = dict `matchWith` selector
+        newDict       = foldl (addShapes stmts) dict $ mapsAndSpecs
```
Comment (hypotext) [inline src/Style.hs]: $ not needed here and on 404

File: src/Style.hs [hunk 4]
```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
...
+    let keys          = M.keys dict
+        identityMap   = M.fromList $ zip keys keys
+        newObjFns     = concatMap (procObjFn identityMap [])    stmts
+        newConstrFns  = concatMap (procConstrFn identityMap []) stmts
+    in (dict, objFns ++ newObjFns, constrFns ++ newConstrFns)
+
+-- Case 2: only one selector, no need to compute combinations.
+-- this function will process: (1) objectives, (2) constraints and (3)
+-- assignments
+procBlock (dict, objFns, constrFns) (selector : [], stmts) =
```
Comment (hypotext) [inline src/Style.hs]: give dict a more descriptive name

File: src/Style.hs [hunk 5]
```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
...
-        selectedSpecs = map
-            (\sel -> let styspecs = matchingSpecs dict sel -- defined externally above procblock
-                         varmaps = map (getVarMap sel) styspecs in zip varmaps styspecs) selectors
+-- Gather a list of objective or constraint functions by mapping over mergedMaps
+-- genFns :: (VarMap -> [t] -> Stmt -> [t]) -> VarMap -> [t]
+genFns f stmts varmap = foldl (f varmap) [] stmts
 
+-- | 'procBlock' is called by 'getDictAndFns'. 'getDictAndFns' would fold this function on a list of blocks, a.k.a. a Style program, and accumulate objective/constraint functions, and a dictionary of geometries to be rendered.
+-- TODO: first split `stmts` into assigns, objs, and contrs and call procConstrFn and procObjFn with `map`?
+procBlock :: (Autofloat a) =>
```
Comment (hypotext) [inline src/Style.hs]: what does "proc" mean? consider giving all the "proc" functions more descriptive names (e.g. "blockToIR")

File: src/Style.hs (hunk 3 above)
Comment (hypotext) [inline src/Style.hs]: looks like we’re matching the sub dict with the selector, but then adding shapes to the dict? is it possible that near the beginning of a style program, we might try to match on something that hasn’t been added to the dict yet?

File: src/Style.hs [hunk 6]
```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
...
+-- TEMP: the system does not support assignment statements in these blocks,
+-- because it is not clear that which Substance object the shapes associate with
+procBlock (dict, objFns, constrFns) (selectors, stmts) =
+    let mapsAndSpecs  = dict `matchWithAll` selectors
+        mergedMaps    = mergeMaps $ map (map fst) $ mapsAndSpecs
+        newObjFns     = concatMap (genFns procObjFn stmts)    mergedMaps
+        newConstrFns  = concatMap (genFns procConstrFn stmts) mergedMaps
+    in (dict, objFns ++ newObjFns, constrFns ++ newConstrFns)
+    where
+        validMap :: [(Name, Name)] -> Bool
```
Comment (hypotext) [inline src/Style.hs]: can you add a comment on the purpose of these functions in the where clause?

File: src/Style.hs [hunk 7]
```diff
@@ -283,12 +269,40 @@ attribute = identifier -- TODO: Naming convention - same as identifiers?
...
+-- * a list of geometric primitives
+-- * a list of constraints
+-- * a list of objectives
+
+-- Type aliases for readability in this section
+-- | 'StyContext' maintains the current output of the translater
+type StyContext a =
+    (StyDict,         -- | dictionary mapping Substance ID to Style objects
+    [ObjFnInfo a],    -- | List of objective functions
+    [ConstrFnInfo a]) -- | List of constraints
```
Comment (hypotext) [inline src/Style.hs]: should we also build a list of computations in Style? right now i'm doing it in an inconsistent manner:
* since comps can only set properties of shapes, they are added to a shape's list of properties (config) in addSpec
* comps are discovered and argument names resolved in Runtime, in compsAndVars (called by getShape)

File: src/Style.hs [hunk 8]
```diff
@@ -283,12 +269,40 @@ attribute = identifier -- TODO: Naming convention - same as identifiers?
...
--- Functions used by "Runtime"
+-- Functions used by "Runtime" -- Style Translator
+-- Translates from a Style AST to:
+-- * a list of geometric primitives
+-- * a list of constraints
+-- * a list of objectives
+
+-- Type aliases for readability in this section
+-- | 'StyContext' maintains the current output of the translater
+type StyContext a =
```
Comment (hypotext) [inline src/Style.hs]: this name might be confusing because i refer to a VarMap as a context (and in PL a context is typically a list of `varName : varType` of variables in the current scope)

File: src/Style.hs (hunk 5 above)
Comment (hypotext) [inline src/Style.hs]: this new procBlock is a lot clearer than the old one! thanks1

File: src/Style.hs [hunk 9]
```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
...
 
--- Helper functions for procBlock
-
--- removeDups :: [[(VarMap, StySpec)]]
--- TODO: really a helper function. consider moving to "Utils"
-cartesianProduct :: [[a]] -> [[a]]
-cartesianProduct = foldr f [[]] where f l a = [ x:xs | x <- l, xs <- a ]
+--------------------------------------------------------------------------------
+-- Selector implementation
+-- GOAL: given a list of selectors and all Substance identifiers, output a
```
Comment (hypotext) [inline src/Style.hs]: can you give an example here of a "list of selectors"? and what are the semantics of a list of selectors? is it AND (match all of them) or OR (match at least one of them)?

File: src/Style.hs [hunk 10]
```diff
@@ -283,12 +269,40 @@ attribute = identifier -- TODO: Naming convention - same as identifiers?
...
+type StyDict = M.Map Name StySpec
 
------ Parser for Style design
+-- | Style specification for a particular object declared in Substance (declarations and constraints)
+-- (TODO: maybe this is not the best model, since there is no difference between "primary" and "secondary" shapes)
+-- NOTE: for Substance constraints such as `Subset A B`, the 'spId' will be '_Subset_A_B' and the 'spArgs' will be '['A', 'B']'
+data StySpec = StySpec {
+    spType   :: C.SubType,  -- | The Substance type of the object
+    spId     :: String,     -- | The Substance ID of the object
+    spArgs   :: [String],   -- | the "arguments" following the type.  The idea is to capture @f@, @A@ and @B@ in the case of @Map f A B@, which is needed for pattern matching (TODO: Maybe not the best term here.)
```
Comment (hypotext) [inline src/Style.hs]: do we ever use spArgs? what is it used for?

File: src/Utils.hs [hunk 11]
```diff
@@ -63,12 +52,40 @@ subsetSizeDiff = 10.0
 epsd :: Floating a => a -- to prevent 1/0 (infinity). put it in the denominator
 epsd = 10 ** (-10)
 
+-- TODO: check naming convention of this
+labelName :: String -> String
```
Comment (hypotext) [inline src/Utils.hs]: i changed the internal naming scheme to “subObjName styObjName [label]” (e.g. "A xaxis", "A xaxis label"); see Utils in my branch)

File: src/Style.hs [hunk 12]
```diff
@@ -300,130 +314,145 @@ initSpec = StySpec { spType = C.PointT,
...
+        addShapes :: [Stmt] -> StyDict -> (VarMap, StySpec) -> StyDict
+        addShapes statements dct (varmap, spec) =
+            let newSpec = foldl (procAssign varmap) spec statements
+            in M.insert (spId newSpec) newSpec dct
+
+-- Case 3: multiple selectors, which in the current Style languages means
+-- we would take a subset of the cartesian product among the selectors.
+-- TEMP: the system does not support assignment statements in these blocks,
+-- because it is not clear that which Substance object the shapes associate with
+procBlock (dict, objFns, constrFns) (selectors, stmts) =
```
Comment (hypotext) [inline src/Style.hs]: design Q: do we want to consider other ways to join/combine selectors? e.g. `Set X | Point X { … }` (matches a set or a point) if so, we'd need to come up with use cases for that, and see how the Style compiler would need to generalize

**Comment by hypotext** [comment]
Looks a lot more readable, thanks! General comments:

* Since there are a lot of nested folds, I didn't follow everything that was happening. Can you summarize the algorithm in a couple of sentences? Something like "For each block, for each of its selectors build a variable mapping. Then for each constraint/objective in each block [resolve variables and generate the function]. And for each object in the block, [generate a shape specification] including each of the object's properties." (I'm probably missing layers here)
* Are there any invariants I should know about, and that should be maintained?
* Line 506 FIXME: is the variable resolution still incorrect?

File: src/Style.hs (hunk 3 above)
Comment (wodeni) [reply src/Style.hs]: No. If you look for `initDict`, you will see we add all Substance IDs __before__ the first call of `procBlock`

File: src/Style.hs (hunk 6 above)
Comment (wodeni) [reply src/Style.hs]: `434e2f2` gives a better implementation. Hope it's clearer

File: src/Style.hs (hunk 8 above)
Comment (wodeni) [reply src/Style.hs]: I think things similar to `VarMap` are usually called symbol table. I always thought "context" is a set of environment variables that includes a symbol table. How about `StyEnv` for "Style environment"?

File: src/Style.hs (hunk 9 above)
Comment (wodeni) [reply src/Style.hs]: I did try to cover it the [Style document](https://github.com/wodeni/notes-public/blob/master/penrose/style-compiler-doc.md#pattern-matching). I'll add examples to the doc for better explanation.

File: src/Style.hs (hunk 10 above)
Comment (wodeni) [reply src/Style.hs]: > "The idea is to capture @f@, @A@ and @B@ in the case of @Map f A B@, which is needed for pattern matching (TODO: Maybe not the best term here.)"

So this `spArgs` field exists because some Substance objs are defined by an n-tuple. For instance, `Subset A B` is effectively `(A, B)` with a type `Subset`. Therefore, to pattern match on them, we need a record of all arguments in a tuple.

**Comment by wodeni** [comment]
> "For each block, for each of its selectors build a variable mapping. Then for each constraint/objective in each block [resolve variables and generate the function]. And for each object in the block, [generate a shape specification] including each of the object's properties." 

That looks like all of them. I will add details of it in the Style document.

File: src/Style.hs (hunk 8 above)
Comment (hypotext) [reply src/Style.hs]: i agree that symbol table is a better name for `VarMap`, but not sure about environment, since this type describes something that's more of a "concrete output" and not just "auxiliary information"

**Comment by hypotext** [comment]
> Can you elaborate? I can't think of a particular one yet

There are some invariants in Runtime like "the shapes in the list of objects must always be kept in the same order so packing and unpacking can work correctly." Wondering if there's anything like that in Style.

> It wasn't an accurate comment. The variable resolution is not "incorrect", but was a design choice that might just be necessary. You need to keep holding the VarMap before you get down to an Id, where the resolution actually happen. Therefore, I think it will stay necessary unless we do a preprocessing pass to resolve the ids first. There could be many ways. For example, do pattern matching first and generate a new Style file with all concrete selectors (Set `A`)

Yeah, as the Style language develops we should rework the name resolution (right now I haven't done it consistently for args to computations vs. args to functions in optimization).

**Comment by wodeni** [comment]
@hypotext I don't think we have invariants like that in `Style`, since it depends on the AST, which is structured. The biggest thing is probably that __a Style program is a collection of blocks__ and __a block is a collection of statements, which can be objectives, constraints, or object declarations__. If these two facts change, changes are needed to these `proc*` functions.

There are surely hardcoded components in `procExpr` (since it was hacky to begin with) and let me know if you have questions about that.

I also added a detailed explanation of how multiple selectors work in the [Style doc](https://github.com/wodeni/notes-public/blob/master/penrose/style-compiler-doc.md)

**Comment by hypotext** [comment]
Thanks!
